### DISCUSSION TITLE
Passing Rows of Data from Repeater to App State

### DISCUSSION THREAD

**Discussion by mmalessa80** [Budibase/budibase#6829] [Help]
Many of my forms for new data entry are dependent on relationship data from a detail (/:id) page. In order to have access to some of this data, when navigating to a form page I therefore attempt to store needed data to app state.

In order to get this accomplished, I have a button on my detail page with actions that update State like this:

![image](https://user-images.githubusercontent.com/106758851/180026358-fa8dcbee-f80c-4e53-9716-7868d6dc55f5.png)

However, I have found that rather than having access to the entire data rows from a repeater, it actually only passes this:

![image](https://user-images.githubusercontent.com/106758851/180024582-381b9e42-0c39-417c-839f-96d7124421e6.png)

Then I need to run further queries to get the detailed data from a specific row of data in a relationship table. Is there a way to 'grab' all rows from a repeater or another alternative way? The way I am currently handling this results in minor unwanted lagging between page loads and saving new records.

**Comment by melohagan** [comment]
Hi @mmalessa80 

Repeaters will use the data from each row provided by a data provider. 
If you are using a filter on the *id* in your data provider, then only one row is going to be available.

What you may have to do is have two data providers. One which is filtered, and another that is not.
So:

> Data Provider (not filtered)
>> Data Provider (filtered)
>>> Repeater (pointing to the filtered data provider)

Then you can access the data of the non-filtered provider with the rows binding. E.g. `{{ Non Filtered Data Provider.Rows }}`

Note that will only return the number of rows available as restricted by the limit:
![Screenshot 2022-07-20 at 17 17 14](https://user-images.githubusercontent.com/101575380/180032263-7fb3e3b6-c479-4c8b-8147-4bd31513d140.png)

----

Because of the extra use of a data provider, it may be a good idea to review your current app design and see if there's another way. Maybe another screen could help for example?

**Reply by mmalessa80** [reply]
Maybe I'm not quiet understanding. My detail page has a 'update' form and inside that form I added a table block and set the data component to Repeater.PermitList.MeterInformation. So then I passed that same data from the repeater to app state so I can use it in another page. Changing the set app state to {{ Repeater.PermitList.MeterInformation.Rows }} does not appear to return anything. Looking at my table block considering that it is a relationship, it would appear that I don't have access to the 'Limit' setting but maybe that is because it is a table block vs just a table?

![image](https://user-images.githubusercontent.com/106758851/180038071-fc17b020-a32b-498b-9748-37eea82419c4.png)

![image](https://user-images.githubusercontent.com/106758851/180038288-614bfae3-af06-4739-8cb3-ff7e07832fa1.png)

**Reply by melohagan** [reply]
Hey @mmalessa80 Whenever you use a Table Block, you won't have access to the **Data Provider.Rows** binding because nothing can be nested inside a table block. 

If you use the Data Provider/Table combo, then you'll be able to access the data provider rows binding.

![Screenshot 2022-07-21 at 08 55 02](https://user-images.githubusercontent.com/101575380/180160694-a1dadc93-39bf-42bd-aec9-1a3dee0794f1.png)

![Screenshot 2022-07-21 at 08 55 38](https://user-images.githubusercontent.com/101575380/180160797-a926103e-07f9-4ee9-bca6-09e26428e52d.png)

**Reply by mmalessa80** [reply]
Getting closer, but still not quite there. So I removed my Table Block and instead replaced it with a 'New Data Provider' and inside that I added a 'New Table' and linked the Data to Repeater.PermitList.MeterInformation. So now when it set my state, I do have the 'Data Provider.Rows' binding available, however this links to my current PermitList record, not the related MeterInformation records. What I would need (and does not appear to be available) is 'Data Provider.Rows.0.New Data Provider.Rows', since 'Data Provider' in this case points to PermitList and 'New Data Provider' points to related MeterInformation. Is this possible within Budibase, or do I have to resort to running a query for MeterInformation and then passing the query results to my App State?

![image](https://user-images.githubusercontent.com/106758851/180228026-597dcbdc-e5c1-4908-871d-bdb954e6e4f6.png)

![image](https://user-images.githubusercontent.com/106758851/180228625-97949f5a-7958-4c2d-89ef-4485d77b6eb9.png)

**Reply by melohagan** [reply]
Hey @mmalessa80 

The data providers are separate. So you would never have `Data Provider.Rows.0.New Data Provider.Rows`

Instead you would have the option of either `Data Provider.Rows` or `New Data Provider.Rows`

If you want to get related data, there's an example here: https://docs.budibase.com/docs/fetching-data#displaying-relational-data
